papers/incorrectness24/incorrectness24.tex

```diff
@@ -0,0 +1,366 @@
+We expect it (and other new type checking features) to be available in
+2024. This extended abstract is based on the RFC, but written in
+``Greek letters and horizontal lines'' rather than ``monospaced text''.
+
+\section{Code defects}
+
+The main goal of non-strict mode is to identify defects, but this requires
+deciding what a defect is. Run-time errors are an obvious defect:
+\begin{verbatim}
+  local hi = "hi"
+  print(math.abs(hi))
+\end{verbatim}
+but we also want to catch common mistakes such as mis-spelling a property name,
+even though Luau returns \verb|nil| for missing property accesses.
+For this reason, we consider a larger class of defects:
+\begin{itemize}
+\item run-time errors,
+\item expressions guaranteed to be \verb|nil|, and
+\item writing to a table property that is never read.
+\end{itemize}
+
+\subsection{Run-time errors}
+
+Run-time errors occur due to run-time type mismatches (such as \verb|5("hi")|)
+or incorrect built-in function calls (such as \verb|math.abs("hi")|).
+Precisely identifying run-time errors is undecidable, for example:
+\begin{verbatim}
+  if cond() then
+    math.abs(“hi”)
+  end
+\end{verbatim}
+We cannot be sure that this code produces a run-time
+error, but we do know that if \verb|math.abs("hi")| is executed, it
+will produce an error, so we consider this to be a defect.
+
+\subsection{Expressions guaranteed to be nil}
+
+Luau tables do not error when a missing property is accessed (though embeddings may). So
+\begin{verbatim}
+  local t = { Foo = 5 }
+  local x = t.Fop
+\end{verbatim}
+won’t produce a run-time error, but is more likely than not a
+programmer error. If the programmer intended to
+initialize \verb|x| as \verb|nil|, they could have initialized
+\verb|x = nil|.  For this reason, we consider it a code defect to use
+an expression that the type system infers is of type \verb|nil|, other
+than the \verb|nil| literal.
+
+\subsection{Writing properties that are never read}
+
+There is a matching problem with misspelling properties when writing. For example
+\begin{verbatim}
+  function f()
+    local t = {}
+    t.Foo = 5
+    t.Fop = 7
+    print(t.Foo)
+  end
+\end{verbatim}
+won’t produce a run-time error, but is more likely than not a
+programmer error, since \verb|t.Fop| is written but never read. We can use
+read-only and write-only table properties types for this, and make it an
+error to create a write-only property.
+
+We have to be careful about this though, because if \verb|f| ended
+with \verb|return t|, then it would be a perfectly sensible function
+with type \verb|() -> { Foo: number, Fop: number }|. The only way to detect
+that \verb|Fop| was never read would be whole-program analysis, which is
+prohibitively expensive.
+
+\section{New Non-strict error reporting}
+
+The difficult part of non-strict mode error-reporting is predicting
+run-time errors. We do this using an error-reporting
+pass that synthesizes a type context such that if any of the $x : T$ in
```

Comment:
Could we add the variable used to represent the context here? E.g. 
```
We do this using an error-reporting pass that synthesizes a type context \gamma such that if any of the $x : T$
```

Reply:
Fixed.